replace species dict with `SpeciesInfo` for #392

PythonFZ opened the pull request:
<!-- The issue that is fixed by this PR, if applicable: -->
Fix #392 

#### Summary of additions and changes

- change `species` from `dotdict` to `SpeciesInfo`
- `experiment.species` remains a dictionary so it can be accessed via `species[key]: SpeciesInfo`. There is some redundant information here about the name being the key in the dict and also the `SpeciesInfo.name` but this simplifies the access and prohibits using `[x for x in SpeciesInfo if x.name == key]` all over the place. 
- add `groups` to SpeciesInfo
- replace the dict of `molecules` with `SpeciesInfo`

# TODOs
we should look into that part of the code again and decide what information we should store in SQL and what in HDF5 because there are some redudancies.

PythonFZ committed:
replace species dict with `SpeciesInfo` for #392

PythonFZ committed:
add name to SpeciesInfo

SamTov commented:
Can you elaborate on the redundancies @PythonFZ ?

PythonFZ commented:
> Can you elaborate on the redundancies @PythonFZ ?

Take the species for example. Currently we store information such as charge and mass in the SQL database but they might also exist in the H5MD database so we should decide where we want to store them.

SamTov commented:
> > Can you elaborate on the redundancies @PythonFZ ?
> 
> Take the species for example. Currently we store information such as charge and mass in the SQL database but they might also exist in the H5MD database so we should decide where we want to store them.

I agree we should consider alternatives. In principle though it is in hdf5 if it changes w.r.t time and SQL otherwise. In the case of charge it can be either and therefore can exist in both.

PythonFZ commented:
> > > Can you elaborate on the redundancies @PythonFZ ?
> > 
> > 
> > Take the species for example. Currently we store information such as charge and mass in the SQL database but they might also exist in the H5MD database so we should decide where we want to store them.
> 
> I agree we should consider alternatives. In principle though it is in hdf5 if it changes w.r.t time and SQL otherwise. In the case of charge it can be either and therefore can exist in both.

I do think that we should only store it in the HDF5 database. It can also store time-independent data.
But we should discuss this in more detail next week.

SamTov reviewed (approved):
Everything is good, there are some more places where this is broken though as evidenced in the tests. Once they pass it is ready.

PythonFZ committed:
Merge branch 'main' into fix_#392

# Conflicts:
#	mdsuite/database/experiment_database.py

PythonFZ committed:
fix tests

* add an ValueError if species is not a dict

PythonFZ committed:
convert molecules to SpeciesInfo

* add groups attribute to SpeciesInfo
* replace all getitem access with getattr access for n_particles
* replace data_dict with getitem in test

PythonFZ committed:
remove groups from SpeciesInfo and add MoleculeInfo instead

* class MoleculeInfo(SpeciesInfo)

PythonFZ committed:
fix notebook (access MoleculeInfo)

PythonFZ commented on mdsuite/database/simulation_database.py:
```diff
@@ -97,6 +97,23 @@ def __eq__(self, other):
+    groups: dict
+        A molecule specific dictionary for mapping the molecule to the
+        particles. TODO SamTov can you explain a bit more because I haven't found
+        a good docstring to copy/paste from map_molecules.py
```
@SamTov could you update the docstrings here?

PythonFZ replied:
+ update notebook

SamTov replied:
Resolved in [1aaa880](https://github.com/zincware/MDSuite/pull/497/commits/1aaa880c3a1bdeff4e715b2912d089de64877d33)

PythonFZ committed:
fix test

- add __eq__ in MoleculeInfo
- add MoleculeInfo to the test
- remove assertDeepAlmostEqual
- add round(x, 14) to molecular_mass

PythonFZ committed:
fix tests

SamTov committed:
run black and isort
